### DISCUSSION TITLE
Unclear cookie behavior in middleware

### DISCUSSION THREAD

**Discussion by Lepiloff** [fastapi/fastapi#6334] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
middleware.py

async def add_user_data_to_request(request: Request, call_next):
    response = await call_next(request)
    user = await is_authenticated(request)
    if user:
        # add user_id to response
        response.set_cookie(
            key="user_id",
            value=user,
            httponly=True,
            max_age=1800,
            expires=60,
        )
        notifications = await MessagesService.get_count(user)
        print(f'notifications: {notifications}')
        if notifications:
            print('not')
            # add notification to response
            response.set_cookie(
                key="event_notifications",
                value=notifications['events'],
                httponly=True,
                max_age=1,
                expires=1,
            ) if notifications['events'] else None
    return response


template.html

{% if request.cookies.event_notifications %}
<small class="notification-badge">{{request.cookies.event_notifications}}</small>
{% else %}
<small class="notification-badge-empty">0</small>
{% endif %}
```

### Description

I am trying to add a user notification system. To do this I add the values I need to cookies using middleware. Using static html template jinja2 to interact with the user. 
The way I see how this should work 
Every time the user refreshes the page in the browser, middleware sends a request to the database to get the message count.If the counter has a non-zero value, a new value is added to the cookie using `response.set_cookie`. The html template checks if this value exists, and if it does, it is output in the interface. 

The issue

So I send a message to a user, so `notifications = await MessagesService.get_count(user) `returns a response to me like this 
`{'total': 1, 'events': 1}`
and I can see it in the console if I `print(f'notifications: {notifications}')`

But in the template, the value of the message counter is not displayed. But what is interesting and completely confusing is that if I refresh the page twice in one second, the counter starts to be displayed. Apparently this is due to the value of `max_age=1`. For experiment I tried to increase `max_age = 3` and now that the counter is displayed twice to reload the page for 3 seconds. If I reload the page every 4 seconds or more, nothing is displayed. 

### Operating System

Linux

### FastAPI Version

0.73.0

### Python Version

3.9

### Additional Context

p/s 
I may not have chosen the most appropriate way to implement the notification system. If you have any ideas on how to improve it, I would be very interested to hear about it.

**Comment by rafsaf** [comment]
Basically you need 2 things: 
- deliver information from server to a user that he has new notification(s)
- deliver information from user to a server that he saw it and we should stop inform him about it.

With that in mind, cookies are not the best way to go in my opinion, and `expires=1` or something like that makes sense only when you want notification alert based on time, which is not usually a case, you user can visit a site after 2 days and want to see notification. And you cant really handle second requirement in that setup where server doesn't care about client and always send this cookies. Except you reset number of notification **every time** it is more than 0.

Some standard way is to have some kind of counter with "number of notification" that is unique for every user, for example stored as field in database in user table. And you already have something that look like that, `notifications = await MessagesService.get_count(user)`. Now first requirement is simple, you can use cookies (why not), i would use header but this is my taste or even separate GET endpoint where i ask for a number of notifications.

And then your client need a way to reset his "number of notification" field in user table, for example you can make POST request **when user click on notification div** to a server or something like that and this action set it to 0, in next reload there will be no notifications.

**Comment by Lepiloff** [comment]
@rafsaf thanks for reply!
I may not have understood you exactly, but I have it  as you described. There is a table in the database where request `notifications = await MessagesService.get_count(user)`. Every time the user reloads the page this request is executed.  There is a functionality which resets this table in the database at the moment of reading the message. But it is not clear why the displaying of the notification counter works so strangely.

**Comment by tiangolo** [comment]
Thanks for the help @rafsaf ! ☕

I assume you solved it, thanks for closing the issue @Lepiloff  👍
